lib/gui/app/models/files.js

```diff
@@ -16,133 +16,92 @@
-const fs = Bluebird.promisifyAll(require('fs'))
 
 /* eslint-disable lodash/prefer-lodash-method */
+/* eslint-disable no-undefined */
+
+const CONCURRENCY = 10
+
+const collator = new Intl.Collator(undefined, {
+  sensitivity: 'case'
+})
 
 /**
- * @summary Async exists function
- * @function
- * @public
- *
- * @description
- * This is a promise for convenience, as it never fails with an exception/catch.
- *
- * @param {String} fullpath - full path
- * @returns {Boolean}
+ * @summary Sort files by their names / stats
+ * @param {FileEntry} fileA - first file
+ * @param {FileEntry} fileB - second file
+ * @returns {Number}
  *
  * @example
- * files.existsAsync('/home/user/file').then(console.log)
- * > true
+ * files.readdirAsync(dirname).then((files) => {
+ *   return files.sort(sortFiles)
+ * })
  */
-exports.existsAsync = (fullpath) => {
-  return new Bluebird((resolve, reject) => {
-    return fs.accessAsync(fullpath, fs.constants.F_OK).then(() => {
-      resolve(true)
-    }).catch(() => {
-      resolve(false)
-    })
-  })
+const sortFiles = (fileA, fileB) => {
+  return (fileB.isDirectory - fileA.isDirectory) ||
+    collator.compare(fileA.basename, fileB.basename)
 }
 
 /**
- * @summary Get file metadata
- * @function
- * @private
- *
- * @param {String} dirname - directory name
- * @param {String} [basename] - custom basename to append
- * @returns {Object} file metadata
- *
- * @example
- * try {
- *   const file = files.getFileMetadataSync('/home/user')
- *   console.log(`Is ${file.basename} a directory? ${file.isDirectory}`)
- * } catch (error) {
- *   console.error(error)
- * }
+ * @summary FileEntry struct
+ * @class
+ * @type {FileEntry}
  */
-exports.getFileMetadataSync = (dirname, basename = '') => {
-  // TODO(Shou): use path.parse object information here
-  const fullpath = path.join(dirname, basename)
-  const pathObj = path.parse(fullpath)
-
-  // TODO(Shou): this is not true for Windows, figure out Windows hidden files
-  const isHidden = pathObj.base.startsWith('.')
-  const stats = fs.statSync(fullpath)
-
-  return {
-    basename: pathObj.base,
-    dirname: pathObj.dir,
-    fullpath,
-    extension: pathObj.ext,
-    name: pathObj.name,
-    isDirectory: stats.isDirectory(),
-    isHidden,
-    size: stats.size
+class FileEntry {
+  /**
+   * @summary FileEntry
+   * @param {String} filename - filename
+   * @param {fs.Stats} stats - stats
+   *
+   * @example
+   * new FileEntry(filename, stats)
+   */
+  constructor (filename, stats) {
+    const components = path.parse(filename)
+
+    this.path = filename
+    this.dirname = components.dir
+    this.basename = components.base
+    this.name = components.name
+    this.ext = components.ext
+    this.isHidden = components.name.startsWith('.')
+    this.isFile = stats.isFile()
+    this.isDirectory = stats.isDirectory()
+    this.isBlockDevice = stats.isBlockDevice()
+    this.isCharacterDevice = stats.isCharacterDevice()
+    this.stats = stats
```

Comment:
the only thing left in `stats` that we use is `size`. Why not copy it like the rest and remove `FileEntry.stats`?

Reply:
Yeah true, done